File: app/src/main/java/com/kazakago/cleanarchitecture/di/module/DomainModule.java
package com.kazakago.cleanarchitecture.di.module;

import com.kazakago.cleanarchitecture.domain.repository.AboutRepository;
import com.kazakago.cleanarchitecture.domain.repository.CityRepository;
import com.kazakago.cleanarchitecture.domain.repository.WeatherRepository;
import com.kazakago.cleanarchitecture.domain.usecase.AboutUseCase;
import com.kazakago.cleanarchitecture.domain.usecase.AboutUseCaseImpl;
import com.kazakago.cleanarchitecture.domain.usecase.CityUseCase;
import com.kazakago.cleanarchitecture.domain.usecase.CityUseCaseImpl;
import com.kazakago.cleanarchitecture.domain.usecase.WeatherUseCase;
import com.kazakago.cleanarchitecture.domain.usecase.WeatherUseCaseImpl;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class DomainModule {

    @Provides
    @Singleton
    public CityUseCase provideCityUseCase(CityRepository cityRepository) {
        return new CityUseCaseImpl(cityRepository);
    }

    @Provides
    @Singleton
    public WeatherUseCase provideWeatherUseCase(WeatherRepository weatherRepository) {
        return new WeatherUseCaseImpl(weatherRepository);
    }

    @Provides
    @Singleton
    public AboutUseCase provideAboutUseCase(AboutRepository aboutRepository) {
        return new AboutUseCaseImpl(aboutRepository);
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/di/module/DomainModule.kt
package com.kazakago.cleanarchitecture.di.module

import com.kazakago.cleanarchitecture.domain.repository.AppInfoRepository
import com.kazakago.cleanarchitecture.domain.repository.CityRepository
import com.kazakago.cleanarchitecture.domain.repository.WeatherApiRepository
import com.kazakago.cleanarchitecture.domain.repository.WeatherRepository
import com.kazakago.cleanarchitecture.domain.usecase.appInfo.*
import com.kazakago.cleanarchitecture.domain.usecase.city.GetCityUseCase
import com.kazakago.cleanarchitecture.domain.usecase.city.GetCityUseCaseImpl
import com.kazakago.cleanarchitecture.domain.usecase.weather.GetWeatherUseCase
import com.kazakago.cleanarchitecture.domain.usecase.weather.GetWeatherUseCaseImpl
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class DomainModule {

    @Provides
    @Singleton
    fun provideGetWeatherUseCase(weatherApiRepository: WeatherApiRepository, weatherRepository: WeatherRepository): GetWeatherUseCase {
        return GetWeatherUseCaseImpl(weatherApiRepository = weatherApiRepository, weatherRepository = weatherRepository)
    }

    @Provides
    @Singleton
    fun provideGetCityUseCase(cityRepository: CityRepository): GetCityUseCase {
        return GetCityUseCaseImpl(cityRepository = cityRepository)
    }

    @Provides
    @Singleton
    fun provideGetAppVersionUseCase(appInfoRepository: AppInfoRepository): GetAppVersionUseCase {
        return GetAppVersionUseCaseImpl(appInfoRepository = appInfoRepository)
    }

    @Provides
    @Singleton
    fun provideGetMailAddressUrlUseCase(appInfoRepository: AppInfoRepository): GetMailAddressUrlUseCase {
        return GetMailAddressUrlUseCaseImpl(appInfoRepository = appInfoRepository)
    }

    @Provides
    @Singleton
    fun provideGetOfficialSiteUrlUseCase(appInfoRepository: AppInfoRepository): GetOfficialSiteUrlUseCase {
        return GetOfficialSiteUrlUseCaseImpl(appInfoRepository = appInfoRepository)
    }

    @Provides
    @Singleton
    fun provideGetPlayStoreUrlUseCase(appInfoRepository: AppInfoRepository): GetPlayStoreUrlUseCase {
        return GetPlayStoreUrlUseCaseImpl(appInfoRepository = appInfoRepository)
    }

}